crates/client/data-availability/src/lib.rs

```diff
@@ -81,53 +87,87 @@ where
+                };
+
+                if child_key == twox_128(STARKNET_NONCE) {
+                    // collect nonce information in state diff
+                    state_diff
+                        .nonces
+                        .insert(ContractAddress(bytes_to_key(&rest_key)), Nonce(bytes_to_felt(&storage_val)));
+                    num_addrs_accessed += 1;
+                } else if child_key == twox_128(STARKNET_STORAGE) {
+                    // collect storage update information in state diff
+                    if rest_key.len() > 32 {
+                        let (addr, key) = rest_key.split_at(32);
+                        let (addr, key) = (bytes_to_key(addr), bytes_to_key(key));
+
+                        state_diff
+                            .storage_diffs
+                            .entry(ContractAddress(addr))
+                            .and_modify(|v| {
+                                v.insert(StorageKey(key), bytes_to_felt(&storage_val));
+                            })
+                            .or_insert(IndexMap::from([(StorageKey(key), bytes_to_felt(&storage_val))]));
+                        num_addrs_accessed += 1;
                     }
-                }
-
-                if prefix == *SN_STORAGE_PREFIX {
-                    if let Some(data) = event.2 {
-                        // first 32 bytes = contract address, second 32 bytes = storage variable
-                        let write_split = key.split_at(32);
-
-                        storage_diffs
-                            .entry(write_split.0)
-                            .and_modify(|v| v.push((write_split.1, data.0.as_slice())))
-                            .or_insert(vec![(write_split.1, data.0.as_slice())]);
+                } else if child_key == twox_128(STARKNET_CONTRACT_CLASS) {
+                    // collect declared class information in state diff
+                    state_diff
+                        .declared_classes
+                        .insert(ClassHash(bytes_to_felt(&rest_key)), CompiledClassHash(bytes_to_felt(&storage_val)));
+                } else if child_key == twox_128(STARKNET_CONTRACT_CLASS_HASH) {
+                    // collect deployed contract information in state diff
+                    let runtime_api = client.runtime_api();
+                    let class_hash = ClassHash(bytes_to_felt(&storage_val));
+                    let current_block_hash = client.info().best_hash;
+
+                    let contract_exists =
+                        runtime_api.contract_class_by_class_hash(current_block_hash, class_hash).is_ok();
+
+                    if contract_exists {
+                        state_diff.replaced_classes.insert(ContractAddress(bytes_to_key(&rest_key)), class_hash);
+                    } else {
+                        state_diff.deployed_contracts.insert(ContractAddress(bytes_to_key(&rest_key)), class_hash);
                     }
+                    num_addrs_accessed += 1;
                 }
```

Comment:
Now that we have the `crates/client/commitment-state-diff` most of this logic can be removed as this service simply because a listener of the state diff channel and then runs the `state_diff_to_calldata` function on it and handles persistence on the node.

Reply:
Right, should we replace `CommitmentStateDiff` by `ThinStateDiff` tho to make this work ?

Reply:
Also in the DA db, keys should now correspond to starknet `BlockHash` directly instead of substrate `B::Hash`

Reply:
@drspacemn All done we should be good now :)